### PR TITLE
nixos/transmission: add dhtBootstrapNodes option

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

If you are running both the Magnetico crawler and Transmission you really should be bootstrapping one from the other, because they aren't necessarily going to find each other.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ehmry** [comment]
So stupid it's not even wrong.
```
error: option services.transmission.dhtBootstrapNodes description uses DocBook
error: option services.transmission.dhtBootstrapNodes.*.host description uses DocBook
error: option services.transmission.dhtBootstrapNodes.*.port description uses DocBook
...
>   example.enable = mkEnableOption (lib.mdDoc ''your thing'');
```

File: nixos/modules/services/torrent/transmission.nix [hunk 1]
```diff
@@ -286,6 +309,13 @@ in
           ${pkgs.jq}/bin/jq --slurp add ${settingsFile} '${cfg.credentialsFile}' |
           install -D -m 600 -o '${cfg.user}' -g '${cfg.group}' /dev/stdin \
            '${cfg.home}/${settingsDir}/settings.json'
+          ${lib.optionalString (cfg.dhtBootstrapNodes != [ ]) ''
```
Comment (rnhmjoj) [inline nixos/modules/services/torrent/transmission.nix:312]: I recommend to do something like this
```
let
  bootstrapFile = pkgs.writeText "dht.bootstrap"
    (lib.concatMapStrings (node: "${node.host} ${toString node.port}\n") cfg.dhtBootstrapNodes)
in
''
  ln -sf ${bootstrapFile} ${cfg.home}/${settingsDir}
''
```
over using a here-doc string. This pattern tends to break in subtle ways due to parameter expansions.

File: nixos/modules/services/torrent/transmission.nix [hunk 2]
```diff
@@ -249,6 +249,29 @@ in
...
+      dhtBootstrapNodes = mkOption {
+        type = with types;
+          listOf (submodule {
+            options = {
+              host = mkOption { type = str; description = "Hostname or IP address."; };
+              port = mkOption { type = port; description = "Port number."; };
+            };
+          });
+        default = [ ];
+        description = "List of nodes from which to bootstap into the DHT.";
```
Comment (ju1m) [inline nixos/modules/services/torrent/transmission.nix:262]: The description should be wrapped by an `mdDoc` nowadays.

**Review by ju1m — CHANGES_REQUESTED** [review]
@ehmry, thanks for your PR, I didn't know about that `dht.bootstrap` file.
Except for that missing `mdDoc`, LGTM.

**Comment by ju1m** [comment]
AFAIU [tr-dht.cc](https://github.com/transmission/transmission/blob/611d36ac84f24adf4b4237b49a411cba3276ca60/libtransmission/tr-dht.cc#L404-L431), creating an empty `dht.bootstrap` file (the default), does not disable the bootstrapping from `dht.transmissionbt.com`. So that's good.

**Comment by ehmry** [comment]
@ju1m I didn't find any documentation on dht.bootstrap, I only found it by grepping.

File: nixos/modules/services/torrent/transmission.nix (hunk 2 above)
Comment (ehmry) [reply nixos/modules/services/torrent/transmission.nix:265]: Those are just strings, not mddoc documents. I'd rather just wait for this "transition period" to pass.

Comment (ju1m) [reply nixos/modules/services/torrent/transmission.nix:265]: @ehmry, yet `mdDoc` is required, have you seen the [manual build failure](https://github.com/NixOS/nixpkgs/actions/runs/3378990469/jobs/5609953849)? Same applies for `host` and `port` options.

Comment (rnhmjoj) [reply nixos/modules/services/torrent/transmission.nix:265]: > yet mdDoc is required

offtopic, but this migration looks crazy to me.

**Comment by AndersonTorres** [comment]
Closing as dead.

**Pull request closed without merging**